src/Microsoft.AspNet.Routing/Template/TemplateMatcher.cs

```diff
@@ -178,6 +179,8 @@ public class TemplateMatcher
 
             TemplatePart parameterNeedsValue = null; // Keeps track of a parameter segment that is pending a value
             TemplatePart lastLiteral = null; // Keeps track of the left-most literal we've encountered
+            TemplatePart firstParameterFound = null;
+            var outValues = new RouteValueDictionary();
```

Comment:
I have a simplifying suggestion here. You can basically leave this implementation untouched and call it twice.

1). Rename `MatchComplexSegment` -> `MatchComplexSegmentCore`
2). Add a new parameter 'int startIndex' to `MatchComplexSegmentCore`
3). Add a new method `MatchComplexSegment` which the same signature as the current code

Logic inside `MatchComplexSegment` (pseudocode):
```
var length = segment.Parts.Length;
if (segment.Parts[length -1].IsOptionalSeparator)
{
     return MatchComplexSegmentCore(..., startIndex: length - 1) || MatchComplextSegmentCore(..., startIndex: length - 3);
}
else
{
    return MatchComplexSegmentCore(..., startIndex: length - 1)
}
```